src/midi/soundcanvas.cpp

```diff
@@ -280,10 +280,88 @@ static void setup_filter(MixerChannelPtr& channel, const bool filter_enabled)
 	}
 }
 
+#if defined(WIN32)
+
+static std::deque<std_fs::path> get_platform_rom_dirs()
+{
+	return {
+	        get_config_dir() / DefaultSoundCanvasRomsDir,
+	};
+}
+
+#elif defined(MACOSX)
+
+static std::deque<std_fs::path> get_platform_rom_dirs()
+{
+	return {
+	        get_config_dir() / DefaultSoundCanvasRomsDir,
+	};
+}
+#else
+
+static std::deque<std_fs::path> get_platform_rom_dirs()
+{
+	// First priority is user-specific data location
+	const auto xdg_data_home = get_xdg_data_home();
+
+	std::deque<std_fs::path> dirs = {
+	        xdg_data_home / "dosbox" / DefaultSoundCanvasRomsDir,
+	        xdg_data_home / "soundcanvas-rom-data",
+	};
+
+	// Second priority are the $XDG_DATA_DIRS
+	for (const auto& data_dir : get_xdg_data_dirs()) {
+		dirs.emplace_back(data_dir / "soundcanvas-rom-data");
+	}
+
+	// Third priority is $XDG_CONF_HOME, for convenience
+	dirs.emplace_back(get_config_dir() / DefaultSoundCanvasRomsDir);
+
+	return dirs;
+}
+
+#endif
+
+static std::deque<std_fs::path> get_rom_dirs()
+{
+	// Get potential ROM directories from the environment and/or system
+	auto rom_dirs = get_platform_rom_dirs();
+
+	// Get the user's configured ROM directory; otherwise use 'mt32-roms'
+	std_fs::path selected_romdir = get_soundcanvas_section()->GetString("soundcanvas_rom_dir");
+
+	if (selected_romdir.empty()) { // already trimmed
+		selected_romdir = DefaultSoundCanvasRomsDir;
+	}
+
+	// Make sure we search the user's configured directory first
+	rom_dirs.emplace_front(resolve_home(selected_romdir.string()));
+	return rom_dirs;
+}
+
 MidiDeviceSoundCanvas::MidiDeviceSoundCanvas()
 {
 	using namespace SoundCanvas;
 
+	// Before attempting to load any plugins, we set the ROM directory location 
+	// hint via the SOUNDCANVAS_ROM_DIR. Note, on POSIX systems 
+	// getenv/setenv are NOT thread safe
+	std::error_code ec;
+	// Search for a directory that exists and is not empty
+	for (auto rom_dir : get_rom_dirs()) {
+		if (std_fs::is_directory(rom_dir, ec)) {
+			if (!std_fs::is_empty(rom_dir, ec) && !ec) {
+				const auto canonical_rom_dir = std_fs::canonical(rom_dir, ec);
+				if (ec) {
+					continue;
+				}
+				LOG_MSG("SOUNDCANVAS: Setting SOUNDCANVAS_ROM_DIR env variable to '%s'", canonical_rom_dir.string().c_str());
+				set_env_var("SOUNDCANVAS_ROM_DIR", canonical_rom_dir.string().c_str());
+				break;
+			}
+		}
+	}
```

Comment:
This isn't very robust. MT-32 actually iterates over every ROM dir rather than stopping at the first non-empty directory.

I tested this with my existing ROM location of `~/.config/dosbox/plugins/Nuked-SC55-Resources` and a simple `touch ~/.config/dosbox/soundcanvas-roms/a.txt` broke my audio. If Windows drops a `thumbs.db` or whatever hidden file it likes to do, people's audio will break.

I would say at the very least, the CLAP plugin should always check `Nuked-SC55-Resources` if it fails to find ROMs in the environment variable location. A better solution would be to handle the actual iteration on the plugin side and maybe we pass a list of directories over in an environment variable.

Reply:
I had a more robust solution but was told to get rid of it because it was too specific...

Personally, I prefer the environment list solution, and beef up the CLAP plugin to handle that case. If I do this, are you guys ok if we merge this as-is, then enhance it with a list solution in a follow-up PR?

Reply:
Although if you're ok to wait a few more days, I can add the list support to the CLAP plugin then implement that here.

Reply:
Why don't you try Staging side iteration through the dirs first and if it's not slow, we're done? In practice, I don't think you'll gain anything by complicating things on the plugin side.

Let's try the simplest thing first 😄

Reply:
Ok, this has been implemented.